### ISSUE TITLE
Javadoc check to avoid unneccesary linebreaks

### ISSUE THREAD

**Issue by SethFalco** [checkstyle/checkstyle#16748]
Would it be fine to have a check included for Javadoc's, that enforces that the full length of a line is used before line breaking?

Example violations, suppose a line length limit of 80.

```java
public class Foobar {

    /**
     * A very long Javadoc string, which will could easily
     * have been written in fewer lines than it was, but for
     * some reason linebreaks were thrown in early. Perhaps
     * because of personal editor preferences, habits, or
     * a lack of awareness/feedback in their tooling.
     */
    public void foo() {
        System.out.println("bar");
    }
}
```

Example fix:

```java
public class Foobar {

    /**
     * A very long Javadoc string, which will could easily have been written in
     * fewer lines than it was, but for some reason linebreaks were thrown in
     * early. Perhaps because of personal editor preferences, habits, or a lack
     * of awareness/feedback in their tooling.
     */
    public void foo() {
        System.out.println("bar");
    }
}
```

In projects I maintain, the rules I informally apply are:

* A paragraph should not contain line breaks until a word would cross the 80th column.
* A paragraph should not cross the 80th column, unless a single word, link, or Javadoc tag us too long to fit.

I'm willing to try to contribute this myself. Just wanted to ask about it first in case I simply missed the check in the documentation, or if this is achievable with a combination of other rules.

**Comment by romani** [comment]
Conceptually I am ok with general idea, but we need to share more details on design and behavior to approve issue and be confident in merge when PR is provided.

Please read https://checkstyle.org/report_issue.html#How_to_request_new_Check.2FModule.3F there are links to examples of how proposal should looks like. Please share all required details, and we will proceed with discussing, approvals and guidance on how to implement it.

**Comment by romani** [comment]
Please share more examples on paragraphs.

**Comment by SethFalco** [comment]
@romani Updated! Thank you for your patience. Is this any better?

I've also replaced the mock example with real ones, except for the `notAViolation1` demo which is a mock since I don't have real-world case for that scenario.

**Comment by romani** [comment]
If we consider as main idea of Check to utilize whole line, to keep comment compact, so it means we should apply it to block tags of javadoc.

```
* @return
     *     Date which the company was converted / closed or dissolved. Refer to
     *     {@link #getCompanyStatus()} to determine which.
```

Should be violated.
If you can share specific reasons to not do this on javadoc block tags, please share it, it will be good reasoning for Check property, to let users choose to validate or not validate block tags.


Please suggest name of Check.

**Comment by SethFalco** [comment]
> Should be violated.

I didn't realize that contradiction while writing this up, as the examples mostly consist of my personal style. Thanks for pointing that out.

> If you can share specific reasons to not do this on javadoc block tags, please share it, it will be good reasoning for Check property, to let users choose to validate or not validate block tags.

I do not have an objective reason not consider it a violation. I do however personally like to have the extra line break on multi-line Javadoc block tags. Just a personal preference!

In my opinion, it would make sense to **not** consider it a violation by default. I'd consider these two separate styles preference:

1. Utilizing the preferred line length on multi-line Javadoc comments.
2. Starting position of (multi-line) Javadoc block tags.

I think the ideal implementation of this would be to disregard the starting position of Javadoc block tags, but continue to enforce the preferred line length wherever it does start.

So both of the following would be valid:

```
     * @return Date which the company was converted / closed or dissolved. Refer
     * to {@link #getCompanyStatus()} to determine which.
```

```
     * @return
     *     Date which the company was converted / closed or dissolved. Refer to
     *     {@link #getCompanyStatus()} to determine which.
```

I think enforcing the starting position of a Javadoc block tag would be a separate Check altogether, which could be used in conjunction with this one to make Javadocs even more compact.

**Comment by SethFalco** [comment]
I've amended the original issue, and believe it encapsulates everything we've discussed.

I did however notice one flaw with my original description of the Check. I described that early line breaks should not be permitted in Javadoc description lines, _including HTML_. 

When I wrote this, I was namely thinking of HTML nodes like `<p>`, `<strong>`, or `<em>` tags. However, it would be very strange to not line break after a closing `</li>` node, imo. I just realized that in my original issue, the `<li>` example perfectly fit the line length limit by coincidence, so missed that I typically do an early line break there.

This falls under formatting, similar to how we're allowing leading line breaks or spaces for Javadoc block tags too. We should probably treat opening and closing HTML nodes the same.

So all of the following would be valid:

```java
    /**
     * Company number stored by Companies House.
     *
     * <p>Is it stored as a string as the company number can begin with letters
     * or padded zeros.</p>
     *
     * @return Company number of the registered company.
     */
```

```java
    /**
     * Company number stored by Companies House.
     *
     * <p>
     * Is it stored as a string as the company number can begin with letters or
     * padded zeros.
     * </p>
     *
     * @return Company number of the registered company.
     */
```

```java
    /**
     * Company number stored by Companies House.
     *
     * <p>
     *     Is it stored as a string as the company number can begin with letters
     *     or padded zeros.
     * </p>
     *
     * @return Company number of the registered company.
     */
```

Does this sound fine to you, to treat opening/closing HTML nodes similar to Javadoc block tags, where we only enforce lines that contain the text content of the node, and to be more permissive with line breaks in such cases? If so, I can amend the issue accordingly.

**Comment by romani** [comment]
Is snippet 2 and 3 are good for Apache Common?

Better to say lines that  starts with html tags will be ignored, as it might imply user formatting. But if any tag in middle of text , it will be under validation.

**Comment by SethFalco** [comment]
> Is snippet 2 and 3 are good for Apache Common?

Last time I worked in Commons BeanUtils a few months ago, it was all over the place tbh. It jumped between the formatting of all 3 snippets.

I actually just looked now, and can see in September 2024 that Gary Gregory addressed the formatting to consistently use snippet 2.

Reference: https://github.com/apache/commons-beanutils/commit/2aa8770dd02aea09fbffc78681a6d6151ce8f606

However, it varies across Apache projects in general. For example, in Apache DeltaSpike, it jumps between snippet 1 and 2, except for when they use the `<pre>` tag, then they do snippet 3.

> Better to say lines that starts with html tags will be ignored, as it might imply user formatting. But if any tag in middle of text , it will be under validation.

Yeah sure, that makes sense to me! I've amended the docs to include this.

I believe I'm satisfied with the proposal now, but let me know if you think anything else needs to be clarified.

**Comment by romani** [comment]
> Multiline Javadocs should anchor to column {0}


Sorry, I do not understand this message. 

what if use somethig like `Text in javadoc line should utilize whole line, some words from next line should be moved to this line`
or any other idea from you.

---------


please move all 1-3 snippets to issue decription .

------


please use this formatting for violations:

https://github.com/checkstyle/checkstyle/blob/720de881c200cf18390c1729cb75829b28a9a161/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethod2.java#L11-L19

it very explicit on location of violation and message so easy to spot nonsense, and you will reuse it as is in tests :)


------

please try to make samples of code compact to avoid 
```
       // May warrant discussion, should Javadoc tags or URLs be required to
        // adhere to the rules like other text? I personally find forcing
        // Javadoc tags and URLs to adhere to the same rules impede on the
        // readability of docs. I personally allow these to to go over the
        // preferred length. If it is at the end of a sentence, then the final
        // period is allowed too. If it is not the end of a sentence, then the
        // sentence should be continued on the next line. This could be a
        // configurable option for the check.

```

you can move them to implementation notes.
URL and any inline tags will be excluded as they might have problems with multiline, but we can decide during massive testing, we can decide later on.

**Comment by SethFalco** [comment]
> Text in javadoc line should utilize whole line, some words from next line should be moved to this line

I think it's important to reference the column number or length in the message. For example:

```
Text in Javadoc line should utilize around 80 characters. Some words from next line should be moved to this line.

Text in Javadoc line should line break around column 80. Some words from next line should be moved to this line.

Line in Javadoc should utilize around 80 characters. Some words from next line should be moved to this line.
```

Without this, it wouldn't be very clear what the fix actually is without checking the Checkstyle config. (When someone contributes to the project.)

Is it fine for the same check to have 2 different violation messages? Since there are 2 scenarios that something can be considered a violation.

1. It line breaks early, cutting lines short.
2. It doesn't line break enough, allowing lines to be too long.

Both messages can share the same first sentence, but then the second clause with the fix would need to change:

```
Text in Javadoc line should utilize around 80 characters. Some words from this line should be moved to a new line.

Text in Javadoc line should utilize around 80 characters. Some words from this line should be moved to the next line.
```

> please move all 1-3 snippets to issue decription .

Done!

> please use this formatting for violations:

Done!

> please try to make samples of code compact to avoid

Done!

---

I also added yet another exception to the list of lines to ignore. I believe we should ignore _all_ content of a `<pre>` tag if it is found in a Javadoc. The `<pre>` is for preformatted text and often snippets of Java code, so we should not enforce or suggest changes to it which would inevitably impact the formatting.

**Comment by romani** [comment]
```
    // violation 8 lines below 'Text in Javadoc line should utilize around 80 characters. Some words from this line should be moved to a new line'
    /**
     * Bootstraps and initializes Commandler application, and pulls in your
     * configuration and runtime dependencies.
     *
     * <p>There are two ways to configure Commandler:</p>
     * 
     * <ul>
     * <li><strong>Static Configuration:</strong> Uses configuration files either in the classpath or externally, and {@link Annotation}s.</li>
     * <li><strong>Dependency Injection Modules:</strong> This entails overriding runtime dependencies for the CDI/IoC container to use.</li>
     * </ul>
     *
     * @author seth@falco.fun (Seth Falco)
     */
    public void violation3() {}
```


this violation looks weird.
`around 80 characters.`  ... but we will strictly enforce 80(lineLimit).
I think we can just do https://github.com/checkstyle/checkstyle/blob/b24e789b51a4b2b0c5486c4bf98d1a30f25d4a37/src/main/resources/com/puppycrawl/tools/checkstyle/checks/sizes/messages.properties#L6 user will figureout how to wrap.

**Comment by SethFalco** [comment]
> Do you have any other reason to let javadoc block tags to not follow general limit ?

For clarity, that was referring to the following scenario, I've marked column 80 with pipes (`|`) just for the visual cue.

```java
    /**                                                                        |
     * Javadoc that links to things like                                       |
     * {@link CompaniesHouse#getRegisteredOfficeAddress(String)}.              |
     *                                                                         |
     * <p>Refer to the properties                                              |
     * {@link RegisteredOfficeAddress#getAddressLine1()} and                   |
     * {@link RegisteredOfficeAddress#getAddressLine2()}.</p>                  |
     *                                                                         |
     * <p>Instead of that, you could invoke                                    |
     * {@link CompaniesHouse#getCompany(String)} and access                    |
     * {@link Company#getRegisteredOfficeAddress()}</p>                        |
     */
    private String strictlyEncorcingLineLength;

    /**                                                                        |
     * Javadoc that links to things like {@link CompaniesHouse#getRegisteredOfficeAddress(String)}.
     *                                                                         |
     * <p>Refer to the properties {@link RegisteredOfficeAddress#getAddressLine1()}
     * and {@link RegisteredOfficeAddress#getAddressLine2()}.</p>              |
     *                                                                         |
     * <p>Instead of that, you could invoke {@link CompaniesHouse#getCompany(String)}
     * and access {@link Company#getRegisteredOfficeAddress()}</p>             |
     */
    private String ehMaybeWeCanGiveTagsAPass;
```

So not tags like `@return` and `@param`, but for tags like `{@link …}` where sometimes it feels silly/excessive, and I'd gravitate towards a few longer lines over many short lines.

But tbh, I have no problem just always following the the length limit strictly if it was enforced. This happens very rarely, and I don't like how it looks/feels in either case. ^-^'

I'll just remove that bullet point, and we can always enforce it strictly.

So changes made:

* I've removed a reference to exceptions for inline Javadoc tags.
* Updated the violation message on long lines.

**Comment by romani** [comment]
Ok it is called "inlined tag".
But
```
/**                                                                        |
     * Javadoc that links to things like {@link CompaniesHouse#getRegisteredOfficeAddress(String)}.
     *
```

Can be:
```
/**                                                                        |
     * Javadoc that links to things like 
     *  {@link CompaniesHouse#getRegisteredOfficeAddress(String)}.
     *
```

And might be ok only if inlined javadoc tag is starting line and too long to finish on limit. In this case our model is more simple and predictable: inlined tag and URL should start a line to exceed lineLimit.
Is it ok for your style vision?

--------

One more item, is ability of suppression. Unfortunately we don't have xpath suppression. If we put violated line in message users will be able to suppress edge cases. I am not strong here at all just sharing idea.

**Comment by SethFalco** [comment]
Sorry for the slow response!

>  Is it ok for your style vision?

Yes, this makes sense to me. I'd be happy to enforce the rule in this way for projects I maintain.

**Comment by romani** [comment]
I am good with this.
Thanks a lot for hard work to make design!!

@nrmancuso , please do final review to approve.

**Comment by nrmancuso** [comment]
Let’s do it

**Comment by aclfe** [comment]
Hi @romani and @SethFalco,

I would like to pick this issue up if it is currently available.

I’ve reviewed the full discussion history and the requirements. I am confident I can implement `JavadocUtilizingTrailingSpace`

The way I see it, and just to affirm the requirements, logic needs to have sort of a lookahead mechanism.  We're also going to ensure strict line limits (while ignoring `HTML_ELEMENT` nodes like `<pre>` and `BLOCK_TAG` . Also need to lookout for under-utilized lines as mentioned. 

Is there anything else I must be aware of?

**Comment by romani** [comment]
@aclfe , please start working on it. It will be great Check.
